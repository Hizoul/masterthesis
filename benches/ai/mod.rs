pub mod minmax;
pub mod mcts;
pub mod unsafe_mcts;