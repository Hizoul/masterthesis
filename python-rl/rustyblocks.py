#!/usr/bin/env python3
import math
import sys, ctypes
from os.path import abspath, dirname, realpath
from ctypes import c_char_p, c_uint32, c_uint8, c_int8, Structure, POINTER,c_void_p,c_int64,c_int16,c_double,c_uint64,c_char_p
import numpy as np
from numpy.ctypeslib import ndpointer
from PIL import Image


parent_folder = dirname(realpath(__file__)) + "/"
class Tuple(Structure):
    _fields_ = [("placed", c_uint8),("reward", c_double),("done", c_uint8),("winner", c_uint8)]
    def __str__(self):
        return "({},{},{},{})".format(self.placed, self.reward,self.done,self.winner)

class GameFieldS(Structure):
    pass

# libpath = "../target/debug/"
libpath = "../target/release/"
libname = "rustyblocks"
prefix = {'win32': ''}.get(sys.platform, 'lib')
extension = {'darwin': '.dylib', 'win32': '.dll'}.get(sys.platform, '.so')
absolute_lib_path = abspath(parent_folder + libpath + prefix + libname + extension)
print("Loading rustyblocks from ", absolute_lib_path)
rustLib = ctypes.cdll.LoadLibrary(absolute_lib_path)

rustLib.field_new.restype = POINTER(GameFieldS)
rustLib.field_free.argtypes = (POINTER(GameFieldS), )
rustLib.hex_new.restype = POINTER(GameFieldS)
rustLib.field_clone.argtypes = (POINTER(GameFieldS), )
rustLib.field_clone.restype = POINTER(GameFieldS)
rustLib.hex_free.argtypes = (POINTER(GameFieldS), )
rustLib.hex_do_action_with_answer.argtypes = (POINTER(GameFieldS), c_uint8, c_uint8)
rustLib.hex_do_action_with_answer.restype = Tuple
rustLib.hex_to_array.restype = ndpointer(dtype=ctypes.c_uint8, shape=(121,))
rustLib.free_hex_array.argtypes = (ndpointer(dtype=ctypes.c_uint8, shape=(121,)), )
rustLib.hex_reset.argtypes = (POINTER(GameFieldS), )
rustLib.field_to_string.argtypes = (POINTER(GameFieldS), )
rustLib.field_to_string.restype = c_void_p
rustLib.field_to_json_log.argtypes = (POINTER(GameFieldS), )
rustLib.field_to_json_log.restype = c_void_p
rustLib.field_restore_log.argtypes = (POINTER(GameFieldS), c_char_p)
rustLib.field_to_array.argtypes = (POINTER(GameFieldS), c_uint8)
rustLib.field_to_array.restype = ndpointer(dtype=c_int8, shape=(420,))
rustLib.free_field_array.argtypes = (ndpointer(dtype=c_int8, shape=(420,)), )
rustLib.field_get_score.argtypes = (POINTER(GameFieldS),)
rustLib.field_get_score.restype = ndpointer(dtype=c_int16, shape=(2,))
rustLib.free_score_array.argtypes = (ndpointer(dtype=c_int16, shape=(2,)), )
rustLib.field_do_action.argtypes = (POINTER(GameFieldS), c_uint8)
rustLib.field_do_action.restype = c_uint8
rustLib.field_do_action_with_answer.argtypes = (POINTER(GameFieldS), c_uint8, c_uint8)
rustLib.field_do_action_with_answer.restype = Tuple
rustLib.field_do_action_with_answer_heu.argtypes = (POINTER(GameFieldS), c_double,c_double,c_double,c_double, c_uint8)
rustLib.field_do_action_with_answer_heu.restype = Tuple
rustLib.field_do_action_self_play.argtypes = (POINTER(GameFieldS), c_uint8, c_int8)
rustLib.field_do_action_self_play.restype = Tuple
rustLib.field_is_game_over.argtypes = (POINTER(GameFieldS), )
rustLib.field_is_game_over.restype = c_uint8
rustLib.field_get_reward.argtypes = (POINTER(GameFieldS), )
rustLib.field_get_reward.restype = c_double
rustLib.field_counter_action.argtypes = (POINTER(GameFieldS), )
rustLib.field_counter_action.restype = c_int64
rustLib.field_counter_action_index.argtypes = (POINTER(GameFieldS), )
rustLib.field_counter_action_index.restype = c_uint8
rustLib.field_reset.argtypes = (POINTER(GameFieldS), )
rustLib.free_string.argtypes = (c_void_p, )
rustLib.eval_heuristic_weights.argtypes = (c_double,c_double,c_double,c_double)
rustLib.eval_heuristic_weights.restype = c_uint64

def field_to_string(field):
  ptr = rustLib.field_to_string(field)
  try:
    return ctypes.cast(ptr, ctypes.c_char_p).value.decode('utf-8')
  finally:
    rustLib.free_string(ptr)
def field_to_log(field):
  ptr = rustLib.field_to_json_log(field)
  try:
    return ctypes.cast(ptr, ctypes.c_char_p).value.decode('utf-8')
  finally:
    rustLib.free_string(ptr)

def field_to_array(field, self_play_is_second_player):
  original_array = rustLib.field_to_array(field, self_play_is_second_player)
  result = original_array.copy().reshape((42,10))
  rustLib.free_field_array(original_array)
  return result

def field_score(field):
  original_array = rustLib.field_get_score(field)
  result = original_array.copy()
  rustLib.free_score_array(original_array)
  return result

def field_to_cnn_array(field, self_play_is_second_player):
  original_array = rustLib.field_to_array(field, self_play_is_second_player)
  result = original_array.copy()
  newres = np.zeros(shape=(32,32,3), dtype=np.uint8)
  for i in range(0, 200):
    y = math.floor(i / 10)
    x = i % 10
    current = 0
    val = 0
    if result[i] == 1 or result[i] == 2:
      current = 0
      if result[i] == 1:
        val = 128
      else:
        val = 255
    elif result[i] == 3 or result[i] == 4:
      current = 2
      if result[i] == 3:
        val = 128
      else:
        val = 255
    newres[y][x][current] = val
  for i in range(200, 210):
    y = math.floor((i-200) / 10)
    x = (i % 10) + 10
    val = result[i] + 124
    newres[y][x][1] = val
  for i in range(210, 230):
    y = math.floor((i-200) / 10)
    x = (i % 10) + 10
    val = result[i]
    if val == -1:
      val = 0
    elif val > 0:
      val = val * 50
    newres[y][x][1] = val
  for i in range(230, 420):
    y = math.floor((i-200) / 10)
    x = (i % 10) + 10
    val = result[i]
    if val == -1:
      val = 0
    else:
      val = 255
    newres[y][x][1] = val
  rustLib.free_field_array(original_array)
  return newres

def hex_to_array(field):
  original_array = rustLib.hex_to_array(field)
  result = original_array.copy().reshape((11,11))
  rustLib.free_hex_array(original_array)
  return result

def field_string_to_array(field_string):
  field_array = np.zeros((20,10), dtype=np.uint8)
  row = 0
  col = 0
  for i in range(len(field_string)):
    entry = field_string[i]
    if entry == '\n':
      col = 0
      row += 1
    elif row < 20 and col < 10:
      field_array[row][col] = ord(entry) - 48
      col += 1
  return field_array

def save_img(data, name):
  Image.fromarray(data, 'RGB').save(name)