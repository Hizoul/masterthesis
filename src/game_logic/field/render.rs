use super::GameField;
use crate::game_logic::val::{GAMEFIELD_MAX_WIDTH,GAMEFIELD_MAX_WIDTH_USIZE,GAMEFIELD_MAX_HEIGHT_USIZE,GAMEFIELD_MAX_HEIGHT,AMOUNT_OF_UNIQUE_PIECES_USIZE};

pub trait Renderable {
  fn to_field_string(&self) -> String;
  fn to_field_string_with_player(&self, include_current_player: bool) -> String;
  fn to_field_string_with_player_as_u8(&self) -> Vec<u8>;
  fn to_playable_string(&mut self) -> String;
  fn to_reshapeable_array(&mut self, second_player_as_first: bool) -> Vec<i8>;
  fn to_reshaped_array(&mut self) -> Vec<Vec<i8>>;
  fn to_image(&mut self, second_player_as_first: bool) -> Vec<f32>;
  fn converted(&mut self) -> Vec<f32>;
}
pub const MAX: usize = GAMEFIELD_MAX_HEIGHT_USIZE*GAMEFIELD_MAX_WIDTH_USIZE;
pub const MAX_FOR_SCORE: usize = (GAMEFIELD_MAX_HEIGHT_USIZE+1)*GAMEFIELD_MAX_WIDTH_USIZE;
pub const MAX_FOR_BLOCKS: usize = (GAMEFIELD_MAX_HEIGHT_USIZE+3)*GAMEFIELD_MAX_WIDTH_USIZE;
// 1 for scores, 2 and 3 for block amounts of players, 19 more for valid play indicators
pub const AMOUNT_OF_ADDITIONAL_LINES: usize = 22;

impl Renderable for GameField {
  fn to_field_string(&self) -> String {
    self.to_field_string_with_player(false)
  }
  fn to_field_string_with_player(&self, include_current_player: bool) -> String {
    let mut field = String::with_capacity((GAMEFIELD_MAX_WIDTH_USIZE * GAMEFIELD_MAX_HEIGHT_USIZE) +
        GAMEFIELD_MAX_WIDTH_USIZE + 1);
    field.push('\n');
    for ys in 0..GAMEFIELD_MAX_HEIGHT {
      let y = (GAMEFIELD_MAX_HEIGHT - 1) - ys;
      for x in 0..GAMEFIELD_MAX_WIDTH {
        let mut player = "0".to_string();
        'search: for p in self.tetrominos.iter() {
          let mut i = 0;
          for part in self.shape_cache.get_block_shape(p).iter() {
            if part.x == x && part.y == y {
              let to_set = if i == 0 {((p.from + 1) * 2) - 1} else {((p.from + 1) * 2)};
              player = to_set.to_string();
              break 'search;
            }
            i += 1;
          }
        }
        field.push_str(player.as_str());
      }
      field.push('\n');
    }
    if include_current_player {
      field.push_str(self.current_player.to_string().as_str());
    }
    field
  }
  fn to_field_string_with_player_as_u8(&self) -> Vec<u8> {
    let mut field: [u8; 201] = [0; 201];
    for p in self.tetrominos.iter() {
      let mut i = 0;
      for part in self.shape_cache.get_block_shape(p).iter() {
        let to_set = if i == 0 {i+=1;((p.from + 1) * 2) - 1} else {((p.from + 1) * 2)};
        field[((((GAMEFIELD_MAX_HEIGHT_USIZE-1) - part.y as usize) * GAMEFIELD_MAX_WIDTH_USIZE) + part.x as usize)] = to_set as u8;
      }
    }
    field[200] = self.current_player as u8;
    field.to_vec()
  }
  fn to_reshapeable_array(&mut self, second_player_as_first: bool) -> Vec<i8> {
    // first additional line = scores
    // second and third additional lines = blocks per players
    // 190 additional entries for possible combinations
    let mut field_nums: Vec<i8> = Vec::with_capacity((GAMEFIELD_MAX_HEIGHT_USIZE+AMOUNT_OF_ADDITIONAL_LINES) * GAMEFIELD_MAX_WIDTH_USIZE);
    for index in 0..(GAMEFIELD_MAX_HEIGHT_USIZE+AMOUNT_OF_ADDITIONAL_LINES)*GAMEFIELD_MAX_WIDTH_USIZE {
      if index >= MAX {
        if index < MAX_FOR_SCORE {
          let mut corrected_index = index % MAX;
          if corrected_index < (self.players as usize) {
            if second_player_as_first {
              corrected_index = if corrected_index == 0 {1} else {0};
            }
            field_nums.push(self.scores[corrected_index] as i8);
          } else {
            field_nums.push(-1);
          }
        } else if index < MAX_FOR_BLOCKS {
          let mut player = (index % MAX_FOR_SCORE) / AMOUNT_OF_UNIQUE_PIECES_USIZE;
          if player >= (self.players as usize) {
            field_nums.push(-1);
          } else {
            let block = (index % MAX_FOR_SCORE) % AMOUNT_OF_UNIQUE_PIECES_USIZE;
            if second_player_as_first {
              player = if player == 0 {1} else {0};
            }
            field_nums.push(self.piece_counter[player][block]);
          }
        } else {
          field_nums.push(-1);
        }
      } else {
        field_nums.push(0);
      }
    }
    for play in self.possible_plays.iter() {
      field_nums[(MAX_FOR_BLOCKS)+play.global_index as usize] = 1;
    }
    for p in self.tetrominos.iter() {
      let mut i = 0;
      for part in self.shape_cache.get_block_shape(p).iter() {
        let mut from = p.from;
        if second_player_as_first {
          from = if from == 0 {1} else {0};
        }
        let to_set = if i == 0 {
          i += 1;
          ((from + 1) * 2) - 1
        } else {
          ((from + 1) * 2)
        };
        field_nums[((((GAMEFIELD_MAX_HEIGHT_USIZE-1) - part.y as usize) * GAMEFIELD_MAX_WIDTH_USIZE) + part.x as usize)] = to_set;
      }
    }
    // let b = self.current_block;
    // let o = self.current_orientation;
    // let x = self.current_x;
    // let y = self.highest_y(b, o, x);
    // if !self.game_over {
    //   let hover_block = self.shape_cache.get_translated(b, o, x, y);
    //   for part in hover_block.iter() {
    //     field_nums[((((GAMEFIELD_MAX_HEIGHT_USIZE-1) - part.y as usize) * GAMEFIELD_MAX_WIDTH_USIZE) + part.x as usize)] = (self.current_player as u8 + 1) * 2;
    //   }
    // }
    field_nums
  }
  fn to_image(&mut self, second_player_as_first: bool) -> Vec<f32> {
    let mut to_return: Vec<f32> = vec![0.0; 3072];
    let orig = self.to_reshapeable_array(second_player_as_first);
    for i in 0..200 {
      let y = (i as f32 / 10.0).floor() as usize;
      let x = i % 10;
      let mut current = 0;
      let mut val = 0.0;
      if orig[i] == 1 || orig[i] == 2 {
        current = 0;
        val = if orig[i] == 1 {128.0} else {255.0};
      } else if orig[i] == 3 || orig[i] == 4 {
        current = 2;
        val = if orig[i] == 3 {128.0} else {255.0};
      }
      to_return[((y * 32*3) + x*3) + current] = val;
    }
    for i in 200..210 {
      let y = ((i - 200) as f32 / 10.0).floor() as usize;
      let x = (i % 10) + 10;
      to_return[(y*32*3)+x*3+1] = orig[i] as f32 + 124.0;
    }
    for i in 210..230 {
      let y = ((i - 200) as f32 / 10.0).floor() as usize;
      let x = (i % 10) + 10;
      let val = if orig[i] <= 0 {0.0} else {orig[i] as f32 * 50.0};
      to_return[(y*32*3)+x*3+1] = val;
    }
    for i in 230..420 {
      let y = ((i - 200) as f32 / 10.0).floor() as usize;
      let x = (i % 10) + 10;
      let val = if orig[i] == -1 {
         0.0
      } else {
        255.0
      };
      to_return[(y*32*3)+x*3+1] = val;
    }
    to_return
  }
  fn converted(&mut self) -> Vec<f32> {
    let not_yet_reshaped = self.to_reshapeable_array(false);
    let mut converted: Vec<f32> = Vec::new();
    for num in not_yet_reshaped {
      converted.push(num as f32);
    }
    converted
  }
  fn to_reshaped_array(&mut self) -> Vec<Vec<i8>> {
    let not_yet_reshaped = self.to_reshapeable_array(false);
    let mut reshaped = Vec::new();
    let mut current_sub = Vec::new();
    let i = 0;
    for num in not_yet_reshaped {
      current_sub.push(num);
      if i > 0 && i % 10 == 0 {
        reshaped.push(current_sub.clone());
        current_sub.clear();
      }
    }
    reshaped
  }
  fn to_playable_string(&mut self) -> String {
    let mut field_nums: Vec<Vec<i8>> = Vec::with_capacity(GAMEFIELD_MAX_HEIGHT_USIZE);
    for _ in 0..GAMEFIELD_MAX_HEIGHT {
      let mut new_line = Vec::with_capacity(GAMEFIELD_MAX_WIDTH_USIZE);
      for _ in 0..GAMEFIELD_MAX_WIDTH {
        new_line.push(0);
      }
      field_nums.push(new_line);
    }
    for p in self.tetrominos.iter() {
      for part in self.shape_cache.get_block_shape(p).iter() {
        field_nums[(GAMEFIELD_MAX_HEIGHT_USIZE-1) - part.y as usize][part.x as usize] = ((p.from + 1) * 2) - 1;
      }
    }
    let b = self.current_block;
    let o = self.current_orientation;
    let x = self.current_x;
    let y = self.highest_y(b, o, x);
    if !self.game_over {
      let hover_block = self.shape_cache.get_translated(b, o, x, y);
      for part in hover_block.iter() {
        let new_y = (GAMEFIELD_MAX_HEIGHT_USIZE-1) - part.y as usize;
        let new_x = part.x as usize;
        if new_x < GAMEFIELD_MAX_WIDTH_USIZE && new_y < GAMEFIELD_MAX_HEIGHT_USIZE {
          field_nums[new_y][new_x] = (self.current_player + 1) * 2;
        }
      }
    }
    let mut field = String::with_capacity((GAMEFIELD_MAX_WIDTH_USIZE * GAMEFIELD_MAX_HEIGHT_USIZE) +
        GAMEFIELD_MAX_WIDTH_USIZE + 1);
    for top in field_nums.iter() {
      for ele in top.iter() {
        field.push_str(ele.to_string().as_str());
      }
      field.push('\n');
    }
    field
  }
}