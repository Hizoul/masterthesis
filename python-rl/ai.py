import sys
import gym
from stable_baselines.common.policies import MlpPolicy
from stable_baselines import PPO2
from stable_baselines.common.vec_env import DummyVecEnv,VecCheckNan
import numpy as np
from os import listdir
from os.path import isfile, join
from stable_baselines.gail import ExpertDataset
from generate_pretraindata import generate_pretraindata_heuristics
from rustyblocks import rustLib, field_to_array, field_to_log

MAX_FOR_BLOCKS = 230
botToUse = sys.argv[1]
# oldLog = sys.argv[2]
oldLog = "{\"log\":[{\"PayloadRolled\":{\"from\":1,\"block\":1}},{\"PayloadPlaced\":{\"from\":1,\"block\":1,\"orientation\":0,\"x\":0,\"y\":0}},{\"PayloadRolled\":{\"from\":0,\"block\":1}},{\"PayloadPlaced\":{\"from\":0,\"block\":1,\"orientation\":0,\"x\":8,\"y\":0}},{\"PayloadRolled\":{\"from\":1,\"block\":1}},{\"PayloadPlaced\":{\"from\":1,\"block\":1,\"orientation\":0,\"x\":2,\"y\":0}},{\"PayloadRolled\":{\"from\":0,\"block\":0}},{\"PayloadPlaced\":{\"from\":0,\"block\":0,\"orientation\":1,\"x\":3,\"y\":2}},{\"PayloadRolled\":{\"from\":1,\"block\":0}},{\"PayloadPlaced\":{\"from\":1,\"block\":0,\"orientation\":1,\"x\":3,\"y\":6}},{\"PayloadRolled\":{\"from\":0,\"block\":2}},{\"PayloadConsidering\":{\"play_index\":0}}]}"

is_box_space = True
dirname = "D:\\4-System\\rusty\\"
filename="50000_heutistic_pretrain_"
filename += "box" if is_box_space else "discrete"
np.seterr(all='raise')

origEnv = gym.make("rustybox-v0" if is_box_space else "rustydiscrete-v0")

origEnv.max_invalid_tries = 7
env = VecCheckNan(DummyVecEnv([lambda: origEnv]))

# Instantiate the agent
model = PPO2.load("models/ppo2boxbestparam/2e4-30.pkl", env=env)
# model.load("models/pretrain/"+filename)

rustLib.field_restore_log(origEnv.field, oldLog.encode('utf-8'))
obs = field_to_array(origEnv.field)
actions, _states = model.predict(obs)
if is_box_space:
  action = 0
  current_max = -1
  for i in range(0, len(actions)):
    action_probability = actions[i]
    newIndex = MAX_FOR_BLOCKS + i
    action_possible = obs[int(newIndex / 10)][newIndex % 10] == 1
    if action_probability > current_max and action_possible:
      current_max = action_probability
      action = i
    rustLib.field_do_action(origEnv.field, action)
else:
  rustLib.field_do_action(origEnv.field, actions)
print("1;"+field_to_log(origEnv.field))