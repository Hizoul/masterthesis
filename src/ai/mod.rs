use crate::game_logic::field::{GameField};
use crate::game_logic::field::render::{Renderable};
use crate::game_logic::field::helper_structs::{GameRules};
use crate::tree::{Node,ThreadSafeNode};
use std::sync::{Arc,RwLock};
pub mod random;
pub mod minmax;
pub mod alphabeta;
pub mod heuristic;
pub mod python;
use std::thread::sleep;
use rayon::prelude::*;
use std::time::Duration;
use std::collections::HashMap;

pub fn get_score(field: &GameField, player: u8) -> i16 {
  field.scores[player as usize]
}

pub trait Bot: Send + Sync {
  fn make_play(&mut self, field: &GameField, is_second: bool) -> u8;
  fn get_name(&self) -> String;
}
const SLEEP_DURATION: u64 = 2;

pub trait TreeBot: Send + Sync {
  fn get_rules(&self) -> GameRules;
  fn get_player(&self) -> u8;
  fn get_score_type(&self) -> u8;
  fn make_tree_play(&mut self, field: &GameField) -> u8;
  fn get_root(&mut self) -> Arc<RwLock<Node>>;
  fn get_transposition_table(&mut self) -> Arc<RwLock<HashMap<Vec<u8>, ThreadSafeNode>>>;
  fn get_played(&mut self) -> &mut Vec<u8>;
  fn get_score(&self, field: &GameField) -> i16;
  fn find_node(&mut self, played_indices: &Vec<u8>) -> ThreadSafeNode {
    let mut parent_arc = self.get_root();
    let rules = self.get_rules();
    let mut played = self.get_played().clone();
    let transposition_table = self.get_transposition_table();
    played.clear();
    let mut depth = 0;
    for i in played_indices.iter() {
      depth += 1;
      played.push(*i);
      let subpointer = parent_arc.clone();
      let child_is_not_parent: bool;
      let c;
      {
        let n = subpointer.read().unwrap();
        c = n.children[*i as usize].clone();
        let newchild = c.read().unwrap();
        child_is_not_parent = *newchild != *n;
      }
      if child_is_not_parent {
        parent_arc = c;
      } else {
        let mut f = GameField::new_with_rules(2, self.get_rules());
        for to_play in played.iter() {
          f.place_block_using_play(*to_play);
        }

        let rendered_field = f.to_field_string_with_player_as_u8();
        let mut existing_child = Option::None;
        if depth >= 4 {
          let readable_table = transposition_table.read().unwrap();
          let found_child = readable_table.get(&rendered_field);
          if found_child.is_some() {
            existing_child = Option::Some(found_child.unwrap().clone());
          }
        }
        if existing_child.is_some() {
          let new_existing_child = existing_child.unwrap();
          {
            let readable = new_existing_child.read().unwrap();
            let mut sfcomp = GameField::new_with_rules(2, rules);
            for to_play in readable.played_indices.iter() {
              sfcomp.place_block_using_play(*to_play);
            }
          }
          {
            let mut attempt = parent_arc.try_write();
            while attempt.is_err() {
              sleep(Duration::from_nanos(10));
              attempt = parent_arc.try_write();
            }
            let mut n = attempt.unwrap();
            n.children[*i as usize] = new_existing_child.clone();
          }
          parent_arc = new_existing_child;
        } else {
          let new_node = Node::new_safe_score(self.get_score(&f), self.get_player(), Some(parent_arc.clone()), Some(f.possible_plays.len()), played.clone());
          // n.children.push(new_node);
          {
            let mut attempt = parent_arc.try_write();
            while attempt.is_err() {
              sleep(Duration::from_nanos(10));
              attempt = parent_arc.try_write();
            }
            let mut n = attempt.unwrap();
            n.children[*i as usize] = new_node.clone();
          }
          if depth >= 4 {
            let mut write_attempt = transposition_table.try_write();
            while write_attempt.is_err() {
              sleep(Duration::from_nanos(SLEEP_DURATION));
              write_attempt = transposition_table.try_write();
            }
            let mut writeable_table = write_attempt.unwrap();
            let write_res = writeable_table.insert(rendered_field, new_node.clone());
            if write_res.is_some() {
              println!("CONFLICT IN TRANSPOISTION TABLE DETECTED!");
            }
          }
          parent_arc = new_node;
        }

        let new_node = Node::new_safe_score(self.get_score(&f), self.get_player(), Some(parent_arc.clone()), Some(f.possible_plays.len()), played.clone());
        {let mut n = subpointer.write().unwrap();
        n.children[*i as usize] = new_node.clone();}
        parent_arc = new_node.clone();
      }
    }
    parent_arc
  }
  fn expand_node(&mut self, played_indices: &Vec<u8>, parent_arc: ThreadSafeNode) {
    let mut f = GameField::new_with_rules(2, self.get_rules());
    for to_play in played_indices.iter() {
      f.place_block_using_play(*to_play);
    }
    let player = self.get_player();
    let rules = self.get_rules();
    let transposition_table = self.get_transposition_table();
    let child_depth = played_indices.len() + 1;
    (0..f.possible_plays.len()).into_par_iter().for_each(|i| {
      let child_is_parent: bool;
      let c;
      {
        let n = parent_arc.read().unwrap();
        c = &n.children[i];
        let newchild = c.read().unwrap();
        child_is_parent = *newchild == *n;
      }
      if child_is_parent {
        let mut sf = GameField::new_with_rules(2, rules);
        for to_play in played_indices.iter() {
          sf.place_block_using_play(*to_play);
        }
        sf.place_block_using_play(i as u8);

        let rendered_field = sf.to_field_string_with_player_as_u8();
        let mut existing_child = Option::None;
        if child_depth >= 4 {
          let readable_table = transposition_table.read().unwrap();
          let found_child = readable_table.get(&rendered_field);
          if found_child.is_some() {
            existing_child = Option::Some(found_child.unwrap().clone());
          }
        }
        if existing_child.is_some() {
          let new_existing_child = existing_child.unwrap();
          // {
          //   let readable = new_existing_child.read().unwrap();
          //   let mut sfcomp = GameField::new_with_rules(2, rules);
          //   for to_play in readable.played_indices.iter() {
          //     sfcomp.place_block_using_play(*to_play);
          //   }
          //   println!("MATCH IN TRANPSOITION TABLE {} {} {:?} {:?}", sfcomp.to_field_string_with_player_as_u8() == sf.to_field_string_with_player_as_u8(), sf.played_indices == readable.played_indices, sf.played_indices, readable.played_indices);
          // }
          {
            let mut attempt = parent_arc.try_write();
            while attempt.is_err() {
              sleep(Duration::from_nanos(10));
              attempt = parent_arc.try_write();
            }
            let mut n = attempt.unwrap();
            n.children[i] = new_existing_child;
          }
        } else {
          let new_node = Node::new_safe_score(get_score(&sf, player), player, Some(parent_arc.clone()), Some(sf.possible_plays.len()), sf.played_indices.clone());
          // n.children.push(new_node);
          {
            let mut attempt = parent_arc.try_write();
            while attempt.is_err() {
              sleep(Duration::from_nanos(10));
              attempt = parent_arc.try_write();
            }
            let mut n = attempt.unwrap();
            n.children[i] = new_node.clone();
          }
          if child_depth >= 4 {
            let mut write_attempt = transposition_table.try_write();
            while write_attempt.is_err() {
              sleep(Duration::from_nanos(SLEEP_DURATION));
              write_attempt = transposition_table.try_write();
            }
            let mut writeable_table = write_attempt.unwrap();
            let write_res = writeable_table.insert(rendered_field, new_node);
            if write_res.is_some() {
              println!("CONFLICT IN TRANSPOISTION TABLE DETECTED!");
            }
          }
        }
      }
    });
    {
      let mut attempt = parent_arc.try_write();
      while attempt.is_err() {
        sleep(Duration::from_nanos(10));
        attempt = parent_arc.try_write();
      }
      let mut n = attempt.unwrap();
      n.expanded = true;
    }
  }
  fn expand_depth(&mut self, played_indices: &Vec<u8>, parent_arc: ThreadSafeNode, depth: u128, leafs: &mut Vec<ThreadSafeNode>) {
    // let start_time = timestamp();
    if depth > 0 {
      self.expand_node(played_indices, parent_arc.clone());
      let children;
      {
        let parent = parent_arc.read().unwrap();
        children = parent.children.clone();
      }
      let mut play_index = 0;
      for child in children.iter() {
        let mut new_indices = played_indices.clone();
        new_indices.push(play_index);
        self.expand_depth(&new_indices.clone(), child.clone(), depth - 1, leafs);
        play_index += 1;
      }
    } else {
      leafs.push(parent_arc.clone());
    }
    // let end_time = timestamp();
    // let transpos = self.get_transposition_table().read().unwrap().len();
    // println!("DEEPENING to {} took {} seconds found transpositions are {}", depth, (end_time - start_time) / 1000, transpos);
  }
}

impl Bot for dyn TreeBot {
  fn make_play(&mut self, field: &GameField, _: bool) -> u8 {
    self.make_tree_play(field)
  }
  fn get_name(&self) -> String {"treebot".to_owned()}
}