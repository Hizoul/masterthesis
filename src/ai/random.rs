use crate::ai::{Bot};
use crate::game_logic::field::{GameField};
use rand;
use rand::Rng;
use rand_hc::Hc128Rng;
use rand::SeedableRng;
use crate::game_logic::val::{RANDOM_SEED};

pub struct RandomBot {
  rng: Hc128Rng
}

impl RandomBot {
  pub fn new() -> RandomBot {
    let rng: Hc128Rng = Hc128Rng::from_seed(RANDOM_SEED);
    RandomBot {rng}
  }
}

impl Bot for RandomBot {
  fn make_play(&mut self, field: &GameField, _: bool) -> u8 {
    self.rng.gen_range(0, field.possible_plays.len()) as u8
  }
  fn get_name(&self) -> String {"random".to_owned()}
}