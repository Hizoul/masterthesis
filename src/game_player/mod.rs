use crate::game_logic::field::{GameField};
use crate::game_logic::field::helper_structs::{GameRules};
use crate::ai::Bot;
pub mod advantage_checker;
pub mod transposition_amount;
pub mod rating;
pub mod tournament;
pub mod mctsparameval;
pub mod final_eval;

pub struct PlaySupervisor {
  pub players: Vec<Box<dyn Bot>>,
  pub field: GameField
}

impl PlaySupervisor {
  pub fn new(players: Vec<Box<dyn Bot>>) -> PlaySupervisor {
    let field = GameField::new_with_rules(players.len() as i8, GameRules::deterministic());
    PlaySupervisor {players,field}
  }
  pub fn new_with_rules(players: Vec<Box<dyn Bot>>, rules: GameRules) -> PlaySupervisor {
    let field = GameField::new_with_rules(players.len() as i8, rules);
    PlaySupervisor {players,field}
  }
  pub fn play(&mut self) {
    while !self.field.game_over {
      for p in self.players.iter_mut() {
        if !self.field.game_over {
          let play_to_make = p.make_play(&self.field, false);
          self.field.place_block_using_play(play_to_make);
        }
      }
    }
  }
}