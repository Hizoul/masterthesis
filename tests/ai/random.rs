use rustyblocks::ai::random::{RandomBot};
use rustyblocks::game_player::{PlaySupervisor};
use rustyblocks::game_logic::field::render::Renderable;
use insta::{assert_snapshot};

#[test]
fn ai_random() {
  let b1 = RandomBot::new();
  let b2 = RandomBot::new();
  let mut ps = PlaySupervisor::new(vec!(Box::new(b1), Box::new(b2)));
  ps.play();
  assert_snapshot!("field_after_game", ps.field.to_field_string());
}