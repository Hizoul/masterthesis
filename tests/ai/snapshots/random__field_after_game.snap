---
created: "2019-09-30T08:33:19.949369563Z"
creator: insta@0.8.1
source: tests/ai/random.rs
expression: ps.field.to_field_string()
---

4003000044
3444440340
0022442200
0012041200
0004030434
0004222240
0034021210
0001222000
0000024000
0000023400
0000024200
0000012120
0000044000
0000034000
0200020000
2240020000
1434020000
0003010000
0004422000
0000412000

